W-13851546 caching policy (#359)
mulesoft/docs-gateway: gateway/1.5/modules/ROOT/pages/policies-pdk-template-caching-policy.adoc

. Configure the `test/config/api.yaml` as follows:
+
.Expand to view the `yaml` configuration example.
[%collapsible]
====
[source,yaml]
----
# Copyright 2023 Salesforce, Inc. All rights reserved.
---
apiVersion: gateway.mulesoft.com/v1alpha1
kind: ApiInstance
metadata:
  name: ingress-http
spec:
  address: http://0.0.0.0:8081
  services:
    upstream:
      address: http://backend
      routes:
        - config:
            destinationPath: /anything/echo/
  policies:
    - policyRef:
        name: awesome-caching-v1-0-impl
      config:
        max_cached_values: 10
        start_hour: 18
        end_hour: 10
----
====

. Configure a Flex Gateway instance to debug the policy.
+
For more information, see xref:policies-pdk-debug-local.adoc[].

. Run the `run` command to start the Flex Gateway instance:
+
[source,ssh]
----
make run
----
+
NOTE: The `run` command starts a Flex Gateway instance inside a Docker container. By default, Docker containers use UTC as the local time. 

. Send requests to Flex Gateway by using the following command as example:
+
[source,ssh]
----
curl http://127.0.0.1:8081/catalog/1 -H "cache_check: cache_value"
----
+

> If the request is sent during the configured non-business hours, the response contains the original header. If the request is sent during the configured business hours, the response is empty.

Reviewer: Since the backend is an echo server, If you send a second request with a different header value during the configured non-business hours, youll have the  header of the first request since it was cached. if not the header of the second request will be present

= Data Caching Policy Template
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

Use the data caching policy template as an example of how to xref:policies-pdk-configure-features.adoc#share-data-between-workers-and-caching.adoc[configure policy caching].

== Before You Begin

. xref:policies-pdk-prerequisites.adoc[]

== Policy Use Case

> An antique store that buys and sells items needs to track the stock of each item in its collection and who buys it. To track the inventory movement, the company uses an API where each unique path represents a different item.. 

Reviewer 2: ```suggestion
An antique store that buys and sells items needs to track the stock of each item in its catalog and who buys it. To track the inventory movement, the company uses an API where each unique path represents a different item.
```

Reviewer 2: I suggest you use either collection and catalog to be consistent. 
Same with inventory and stock

> During business hours, the catalog changes constantly and caching the available items isn't very useful. Outside of business hours, the stock is frozen, and the catalog can be cached without any concern. Given this, the manager of the antique store's server decides to implement a caching policy that caches the stock during the non-business hours. 

Reviewer 2: ```suggestion
During business hours, the catalog changes constantly and caching the available items isn't very useful. Outside of business hours, the stock is frozen, and the catalog can be cached without any concern. Given this, the antique store's server manager decides to implement a caching policy that caches the stock during the non-business hours. 
```

Reviewer 2: I'm not sure about "isn't very useful" and "Given this" for global language and conciseness.
Could you simplify something like:
Because the store's catalog items in stock change constantly during business hours but are frozen outside of business hours, the server manager decides to implement a caching policy that caches the stock during non-business hours.

> The policy first checks the time of each incoming request. Then, if the request is outside of business hours and if a response for the path has not been previously cached, the policy caches the response. Then the policy returns the cached response for all following requests to the path outside of business hours. When business hours resume, the cache is cleared and is ready to cache new responses during the next period of non-business hours.

Reviewer 2: How about putting this is a more scannable format? Since it's essentially steps, you could put it something like the following:

The server manager implements the caching policy as follows:

. The policy checks the time of each incoming request. 
. If the request is outside of business hours and if a response for the path has not been previously cached, the policy caches the response. 
. The policy returns the cached response for all following requests to the path outside of business hours. 
. When business hours resume, Anypoint Platform clears the cache and is ready to cache new responses during the next period of non-business hours.

To reuse the policy, the amount of cached requests and the non-business hours are configurable. 

> An error in the caching flow should not make a request fail. So, by default, the caching policy does not block requests.  

Reviewer 2: ```suggestion
An error in the caching flow should not make a request fail. By default, the caching policy does not block requests.  
```

. Configure the `test/config/api.yaml` as follows:
+
.Expand to view the `yaml` configuration example.
[%collapsible]
====
[source,yaml]
----
# Copyright 2023 Salesforce, Inc. All rights reserved.
---
apiVersion: gateway.mulesoft.com/v1alpha1
kind: ApiInstance
metadata:
  name: ingress-http
spec:
  address: http://0.0.0.0:8081
  services:
    upstream:
      address: http://backend
      routes:
        - config:
            destinationPath: /anything/echo/
  policies:
    - policyRef:
        name: awesome-caching-v1-0-impl
      config:
        max_cached_values: 10
        start_hour: 18
        end_hour: 10
----
====

. Configure a Flex Gateway instance to debug the policy.
+
For more information, see xref:policies-pdk-debug-local.adoc[].

. Run the `run` command to start the Flex Gateway instance:
+
[source,ssh]
----
make run
----
+
NOTE: The `run` command starts a Flex Gateway instance inside a Docker container. By default, Docker containers use UTC as the local time. 

. Send a request to Flex Gateway:
+
[source,ssh]
----
curl http://127.0.0.1:8081/catalog/1 -H "cache_check: cache_value"
----
+

> NOTE: The upstream service used in the debugging environment included with PDK responds with an echo of the request.

Reviewer 2: Is PDK being used without "the"? Is PDK the official short name?

Author: I'd say PDK is the official short name, but I'm not exactly sure. Let me talk to the PM and then I will update these all universally later.

. Send another request to Flex Gateway changing the included header:
+
[source,ssh]
----
curl http://127.0.0.1:8081/catalog/1 -H "cache_check: cache_value1"
----
+

> If the requests were made outside of business hours, the header `cache_check: cache_value` is included in the response body instead of `cache_check: cache_value1`. If the requests were made inside of business hours, the header `cache_check: cache_value1` is included in the response body.

Reviewer 2: ```suggestion
If requests are made outside of business hours, the header `cache_check: cache_value` is included in the response body instead of `cache_check: cache_value1`. If requests are made inside of business hours, the header `cache_check: cache_value1` is included in the response body.
```

Reviewer 2: Logic: Avoid past tense

> . Adjust the configured non-business hours and the request path to view different responses.

Reviewer 2: ```suggestion
. Change the non-business hours and request path to view different responses.
```

Reviewer 2: Logic: Use the simpler term (change) and simplify the wording. Wouldn't "the configured" be implied, especially in context?
